foyer/general_forcefield.py

```diff
@@ -0,0 +1,417 @@
+            Validator to yield a more descriptive error message.
+            '''
+            warnings.warn('Invalid XML detected. Could not auto-sort topology '
+                          'objects by precedence.')
+
+        # write to temp file
+        temp_file = NamedTemporaryFile(suffix=suffix, delete=False)
+        with open(temp_file.name, 'w') as temp_f:
+            temp_f.write(xml_contents)
+
+        # append temp file name to list
+        preprocessed_files.append(temp_file.name)
+
+    if backend == 'openmm':
+        return preprocessed_files
+    elif backend == 'gmso':
+        # Run through the forcefield XML conversion
+        return forcefield_files
+    else:
+        raise FoyerError('Backend not supported')
+
+class Forcefield(object):
+    """General Forcefield object that can be created by either GMSO Forcefield or OpenMM Forcefield
+
+    Parameters
+    ----------
+    forcefield_files : list of str, optional, default=None
+        List of forcefield files to load.
+    name : str, optional, None
+        Name of a forcefield to load that is packaged within foyer.
+    backend : str, optional, default='openmm'
+        Name of the backend used to store all the Types' information.
+        Can choose between 'openmm' and 'gmso'.
+
+    """
+    def __init__(self, forcefield_files=None, name=None,
+                       validation=True, backend='gmso',
+                       debug=False):
+        self.atomTypeDefinitions = dict()
+        self.atomTypeOverrides = dict()
+        self.atomTypeDesc = dict()
+        self.atomTypeRefs = dict()
+        self.atomTypeClasses = dict()
+        self.atomTypeElements = dict()
+        self._included_forcefields = dict()
+        self.non_element_types = dict()
+        self._version = None
+        self._name = None
+
+        if forcefield_files is not None:
+            if isinstance(forcefield_files, (list, tuple, set)):
+                all_files_to_load = list(forcefield_files)
+            else:
+                all_files_to_load = [forcefield_files]
+
+        if name is not None:
+            try:
+                file = self.included_forcefields[name]
+            except KeyError:
+                raise IOError('Forcefild {} cannot be found.'.format(name))
+            else:
+                all_files_to_load = [file]
+
+        # Preprocessed the input files
+        preprocessed_files = preprocess_forcefield_files(all_files_to_load, backend=backend)
+        if validation:
+            for ff_file_name in preprocessed_files:
+                Validator(ff_file_name, debug)
+
+        # Load in an internal forcefield object depends on given backend
+        if backend == 'gmso':
+            self._parse_gmso(*preprocessed_files)
+        elif backend == 'openmm':
+            self._parse_mm(*preprocessed_files)
+        elif backend == 'openff':
+            self._parse_ff(*preprocessed_files)
```

Comment:
Should these be `self._parse_gmso(preprocessed_files)` without the asterisk? These functions appear to accept a singular list, not the unwrapped, arbitrary number of arguments

Reply:
I will fix that